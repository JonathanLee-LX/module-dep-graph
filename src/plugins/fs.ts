import { existsSync, readFileSync, statSync } from "node:fs";
import { dirname, resolve } from "node:path";
import { Config, UserPlugin } from "../config";
import { ensureExist } from "../helpers";
import { Module } from "../Module";

export interface FsPluginOptions {
  /**
   *
   * default: '.'
   */
  base?: string;
  /**
   * default: {}
   */
  alias?: Record<string, string>;
  /**
   * default: [".js", ".ts", ".jsx", ".tsx"]
   */
  extensions?: string[];
}

export function createFsPlugin(options: FsPluginOptions): UserPlugin {
  const {
    base = ".",
    extensions = [".js", ".ts", ".jsx", ".tsx"],
    alias = {},
  } = options;

  function isBareImport(moduleName: string) {
    const { alias = {} } = options;

    return (
      /^@?[a-zA-Z]+/.test(moduleName) &&
      Object.keys(alias).every((k) => k !== moduleName.split("/")[0])
    );
  }

  return {
    name: "fs",
    load(id: string) {
      if (id.startsWith("@fs")) return null;
      const code = readFileSync(id, "utf-8");
      return code;
    },
    resolve(id: string, config: Config, importer: Module | null) {
      const isBare = isBareImport(id);
      if (isBare) return "@fs/" + id;

      let resolvedId: string;

      const matched = Object.entries(alias).find(
        ([k]) => k.split("/")[0] === k
      );

      if (matched) {
        id = id.replace(matched[0], matched[1]);
      }

      // foo.svg?inline -> foo.svg
      id = id.replace(/\?[a-zA-Z]+/, "");

      if (id.startsWith("/")) {
        // relative from __dirname
        resolvedId = resolve(base, "." + id);
      } else if (id.startsWith(".")) {
        resolvedId = resolve(importer ? dirname(importer.path) : base, id);
      } else {
        throw new Error(`invalid id ${id}`);
      }

      try {
        const stats = statSync(resolvedId);
        if (stats.isDirectory()) {
          resolvedId = resolve(resolvedId, "./index");
        }
      } catch (error) {
        //
      }

      // relative from entry base
      let exist: boolean = existsSync(resolvedId);

      if (!exist) {
        let i = 0;
        while (i < extensions.length) {
          const ext = extensions[i];
          const guessedFilePath = resolvedId + ext;
          exist = existsSync(guessedFilePath);
          if (exist) {
            resolvedId = guessedFilePath;
            break;
          }
          i++;
        }
      }

      if (!exist)
        throw new Error(`cannot find module ${resolvedId} from file system.`);

      ensureExist(resolvedId);

      return resolvedId;
    },
  };
}
