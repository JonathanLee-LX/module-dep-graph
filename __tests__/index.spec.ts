import { readFileSync } from "node:fs";
import { resolve } from "node:path";
import {
  createModuleGraphFromEntry,
  defineConfig,
  doLoad,
  doResolve,
} from "..";

const config = defineConfig({
  base: resolve(__dirname, "../fixtures"),
});

it("createModuleGraph", async () => {
  const graph = await createModuleGraphFromEntry("./foo.ts", config);
  expect(graph.size).toBeGreaterThan(0);
  expect(graph).toMatchSnapshot();
});

it("doResolve", async () => {
  const foo = doResolve("./foo.ts", config, null);
  expect(foo).toBe(resolve(process.cwd(), "./fixtures/foo.ts"));
});

it("doLoad", async () => {
  const code = doLoad(doResolve("./foo.ts", config, null), config);
  expect(code).toBe(
    readFileSync(resolve(process.cwd(), "./fixtures/foo.ts"), "utf-8")
  );
});
