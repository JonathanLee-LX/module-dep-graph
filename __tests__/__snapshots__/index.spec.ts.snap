// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createModuleGraph 1`] = `
Map {
  "/workspaces/codespaces-blank/fixtures/foo.ts" => Module {
    "deps": Set {},
    "exports": [],
    "imports": [
      {
        "a": -1,
        "d": -1,
        "e": 29,
        "n": "./bar",
        "s": 24,
        "se": 30,
        "ss": 0,
      },
    ],
    "path": "/workspaces/codespaces-blank/fixtures/foo.ts",
  },
  "/workspaces/codespaces-blank/fixtures/bar.ts" => Module {
    "deps": Set {
      Module {
        "deps": Set {},
        "exports": [],
        "imports": [
          {
            "a": -1,
            "d": -1,
            "e": 29,
            "n": "./bar",
            "s": 24,
            "se": 30,
            "ss": 0,
          },
        ],
        "path": "/workspaces/codespaces-blank/fixtures/foo.ts",
      },
    },
    "exports": [
      {
        "e": 22,
        "le": 22,
        "ln": "getBar",
        "ls": 16,
        "n": "getBar",
        "s": 16,
      },
    ],
    "imports": [],
    "path": "/workspaces/codespaces-blank/fixtures/bar.ts",
  },
}
`;
